### PR TITLE
Add support for deprecation

### LINKED ISSUE #20704: Add Deprecation support
Implementation issue to track https://github.com/ballerina-platform/ballerina-spec/issues/18

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
$subject. With this PR deprecation is supported for 

- source type
- source object type
- source const
- source function
- source object function

Fixes #20704

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java [hunk 1]
```diff
@@ -478,7 +478,10 @@
     FIELD_NOT_FOUND("field.not.found"),
     OVERLOADED_METHODS("overloaded.method"),
     UNSUPPORTED_PRIMITIVE_TYPE("unsupported.primitive.type.reason"),
-    METHOD_SIGNATURE_DOES_NOT_MATCH("method.signature.not.match")
+    METHOD_SIGNATURE_DOES_NOT_MATCH("method.signature.not.match"),
+    DEPRECATED("deprecated"),
+    DEPRECATION_ANNOTATION_SHOULD_AVAILABLE("deprecation.annotation.should.available"),
+    DEPRECATION_DOCUMENTATION_SHOULD_AVAILABLE("deprecation.documentation.should.available"),
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java]:
```suggestion
    DEPRECATION_ANNOTATION_SHOULD_BE_AVAILABLE("deprecation.annotation.should.available"),
    DEPRECATION_DOCUMENTATION_SHOULD_BE_AVAILABLE("deprecation.documentation.should.available"),
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserBaseListener.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Generated from BallerinaParser.g4 by ANTLR 4.5.3
+// Generated from /home/kavindu/WSO2-GIT/ballerina-lang/compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 by ANTLR 4.5.3
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserBaseListener.java]: Needs to be undone.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserListener.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Generated from BallerinaParser.g4 by ANTLR 4.5.3
+// Generated from /home/kavindu/WSO2-GIT/ballerina-lang/compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 by ANTLR 4.5.3
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserListener.java]: Needs to be undone.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangMarkDownDeprecationDocumentation.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import org.ballerinalang.model.tree.expressions.MarkDownDocumentationDeprecationAttributeNode;
+import org.wso2.ballerinalang.compiler.tree.BLangNodeVisitor;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Represents deprecation documentation node.
+ *
+ * @since 0.981.0
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangMarkDownDeprecationDocumentation.java]:
```suggestion
 * @since 1.2.0
```

Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangMarkDownDeprecationDocumentation.java]: Changed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserListener.java (hunk 3 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserListener.java]: Changed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserBaseListener.java (hunk 2 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/antlr4/BallerinaParserBaseListener.java]: Changed

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java (hunk 1 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java]: Changed

File: langlib/lang.annotations/src/main/ballerina/src/lang.annotations/annotations.bal [hunk 5]
```diff
@@ -33,3 +33,9 @@ type ArgsData record {|
 
 # Defaultable annotation data generated at compile time. This is for internal use.
 annotation ArgsData DefaultableArgs on function;
+
+# An annotation that marks a program element as deprecated.
+#
+# The usage of a deprecated program element is not recommended due to
+# various reasons. Hence, the compiler issues a warning when such an element is used.
+public const annotation deprecated on source type, source object type, source const, source function, source object function;
```
Comment (hasithaa) [inline langlib/lang.annotations/src/main/ballerina/src/lang.annotations/annotations.bal]: Please create an issue for fields.

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 6]
```diff
@@ -1211,3 +1211,12 @@ error.unsupported.primitive.type.reason=\
 
 error.method.signature.not.match=\
   '{ballerinax/java}'METHOD_SIGNATURE_DOES_NOT_MATCH ''{0}''
+
+warning.deprecated=\
+  ''{0}'' is deprecated
+
+error.deprecation.annotation.should.available=\
+  deprecation annotation should be available, since deprecation documentation is available
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: Shall we reword this to something like this?

invalid `Deprecated` usage on a non-deprecated construct; annotation `deprecated` not found.  

@MaryamZi @pubudu91 @rdhananjaya thoughts?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 7]
```diff
@@ -1211,3 +1211,12 @@ error.unsupported.primitive.type.reason=\
...
+
+error.deprecation.documentation.should.available=\
+  deprecation documentation should be available, since deprecation annotation is available
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: ``no `Deprecated` documentation found in deprecated construct``?

@MaryamZi @pubudu91 @rdhananjaya thoughts?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 8]
```diff
@@ -1211,3 +1211,12 @@ error.unsupported.primitive.type.reason=\
 
 error.method.signature.not.match=\
   '{ballerinax/java}'METHOD_SIGNATURE_DOES_NOT_MATCH ''{0}''
+
+warning.deprecated=\
+  ''{0}'' is deprecated
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: ``Usage of ''{0}'' is deprecated``?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BSymbol.java [hunk 9]
```diff
@@ -46,6 +46,7 @@
     public boolean tainted;
     public boolean closure;
     public MarkdownDocAttachment markdownDocumentation;
+    public boolean isDeprecated;
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BSymbol.java]: This implementation will work only with the same module. Because we don't persist this with BALO. Easy wat to implement this makes Deprecated as a part of the symbol flag.

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 6 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Shall we say

``invalid `Deprecated` documentation on a non-deprecated construct; `@deprecated` annotation not found.``

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]:
```
invalid documentation: 'Deprecated' documentation only allowed on constructs annotated as deprecated using the '@deprecated' annotation
```

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]:
```
invalid documentation: 'Deprecated' documentation only allowed on constructs annotated as '@deprecated'
```

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 7 above)
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]:
```
invalid annotation: constructs annotated as '@deprecated' must have an accompanying 'Deprecated' documentation
```

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]:
```
constructs annotated as '@deprecated' must have an accompanying 'Deprecated' documentation
```

File: langlib/lang.annotations/src/main/ballerina/src/lang.annotations/annotations.bal (hunk 5 above)
Comment (KavinduZoysa) [reply langlib/lang.annotations/src/main/ballerina/src/lang.annotations/annotations.bal]: Created https://github.com/ballerina-platform/ballerina-lang/issues/21621

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 6 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Can't we just say

``invalid documentation: 'Deprecated' documentation is only allowed on constructs annotated as '@deprecated'``

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 7 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: ``constructs annotated as '@deprecated' must have 'Deprecated' documentation``

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: "must have" or "must have a"

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: documentation is non-countable? https://ell.stackexchange.com/questions/16580/difference-between-document-and-documentation

I think we should not have an "a".

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BSymbol.java (hunk 9 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BSymbol.java]: Done

**Comment by KavinduZoysa** [comment]
> Please add across module test cases.

Added tests

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 6 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Changed

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 7 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Changed
